File: web/src/domain/brewday/service/brewday.ts
import { request } from '@domain/client'

import type {
  BrewSession,
  BrewSessionInput,
  GravityReading,
  GravityReadingInput,
  SessionStatus,
} from '../entities'

export const fetchSessions = (): Promise<BrewSession[]> =>
  request('brewday_list_sessions', {
    method: 'GET',
    path: '/api/brewday/sessions',
  })

export const fetchSession = (id: string): Promise<BrewSession> =>
  request(
    'brewday_get_session',
    { method: 'GET', path: `/api/brewday/sessions/${id}` },
    { id },
  )

export const createSession = (body: BrewSessionInput): Promise<BrewSession> =>
  request(
    'brewday_create_session',
    { method: 'POST', path: '/api/brewday/sessions' },
    { body },
  )

export const updateSessionStatus = (
  id: string,
  status: SessionStatus,
): Promise<BrewSession> =>
  request(
    'brewday_update_session_status',
    { method: 'PUT', path: `/api/brewday/sessions/${id}/status` },
    { id, body: { status } },
  )

export const deleteSession = (id: string): Promise<void> =>
  request(
    'brewday_delete_session',
    { method: 'DELETE', path: `/api/brewday/sessions/${id}` },
    { id },
  )

export const fetchCompletedSteps = (sessionId: string): Promise<number[]> =>
  request(
    'brewday_list_completed_steps',
    { method: 'GET', path: `/api/brewday/sessions/${sessionId}/steps` },
    { id: sessionId },
  )

export const setStepCompleted = (
  sessionId: string,
  stepId: number,
  completed: boolean,
): Promise<void> =>
  request(
    'brewday_set_step_completed',
    {
      method: 'PUT',
      path: `/api/brewday/sessions/${sessionId}/steps/${stepId}`,
    },
    { id: sessionId, step_id: stepId, body: { completed } },
  )

export const fetchGravityReadings = (
  sessionId: string,
): Promise<GravityReading[]> =>
  request(
    'brewday_list_gravity_readings',
    { method: 'GET', path: `/api/brewday/sessions/${sessionId}/gravity` },
    { id: sessionId },
  )

export const addGravityReading = (
  sessionId: string,
  body: GravityReadingInput,
): Promise<GravityReading> =>
  request(
    'brewday_add_gravity_reading',
    { method: 'POST', path: `/api/brewday/sessions/${sessionId}/gravity` },
    { id: sessionId, body },
  )

export const deleteGravityReading = (id: string): Promise<void> =>
  request(
    'brewday_delete_gravity_reading',
    { method: 'DELETE', path: `/api/brewday/gravity/${id}` },
    { id },
  )

File: desktop/src/commands/brewday.rs
use serde::Deserialize;
use tauri::State;

use mosto_core::brewday::{
    store, BrewSession, BrewSessionInput, GravityReading, GravityReadingInput, SessionStatus,
};

use crate::state::AppState;

#[derive(Deserialize)]
pub struct StatusUpdate {
    status: SessionStatus,
}

#[derive(Deserialize)]
pub struct StepUpdate {
    completed: bool,
}

// Thin translators around `mosto_core::brewday::store` — identical calls to
// the ones `server/src/brewday/handlers.rs` makes, just returning
// `Result<T, String>` (Tauri's error convention) instead of an HTTP response.

#[tauri::command]
pub async fn brewday_list_sessions(state: State<'_, AppState>) -> Result<Vec<BrewSession>, String> {
    store::list_sessions(&state.pool).await.map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn brewday_get_session(
    state: State<'_, AppState>,
    id: String,
) -> Result<BrewSession, String> {
    store::get_session(&state.pool, &id)
        .await
        .map_err(|e| e.to_string())?
        .ok_or_else(|| "session not found".to_string())
}

#[tauri::command]
pub async fn brewday_create_session(
    state: State<'_, AppState>,
    body: BrewSessionInput,
) -> Result<BrewSession, String> {
    store::create_session(&state.pool, body)
        .await
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn brewday_update_session_status(
    state: State<'_, AppState>,
    id: String,
    body: StatusUpdate,
) -> Result<BrewSession, String> {
    store::update_session_status(&state.pool, &id, body.status)
        .await
        .map_err(|e| e.to_string())?
        .ok_or_else(|| "session not found".to_string())
}

#[tauri::command]
pub async fn brewday_delete_session(state: State<'_, AppState>, id: String) -> Result<(), String> {
    let deleted = store::delete_session(&state.pool, &id)
        .await
        .map_err(|e| e.to_string())?;
    if deleted {
        Ok(())
    } else {
        Err("session not found".to_string())
    }
}

#[tauri::command]
pub async fn brewday_list_completed_steps(
    state: State<'_, AppState>,
    id: String,
) -> Result<Vec<i64>, String> {
    store::completed_steps(&state.pool, &id)
        .await
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn brewday_set_step_completed(
    state: State<'_, AppState>,
    id: String,
    step_id: i64,
    body: StepUpdate,
) -> Result<(), String> {
    store::set_step_completed(&state.pool, &id, step_id, body.completed)
        .await
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn brewday_list_gravity_readings(
    state: State<'_, AppState>,
    id: String,
) -> Result<Vec<GravityReading>, String> {
    store::list_gravity_readings(&state.pool, &id)
        .await
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn brewday_add_gravity_reading(
    state: State<'_, AppState>,
    id: String,
    body: GravityReadingInput,
) -> Result<GravityReading, String> {
    store::add_gravity_reading(&state.pool, &id, body)
        .await
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn brewday_delete_gravity_reading(
    state: State<'_, AppState>,
    id: String,
) -> Result<(), String> {
    let deleted = store::delete_gravity_reading(&state.pool, &id)
        .await
        .map_err(|e| e.to_string())?;
    if deleted {
        Ok(())
    } else {
        Err("gravity reading not found".to_string())
    }
}
